### ISSUE TITLE
Haproxy 3.0.4 crashes with all threads stuck=1

### ISSUE THREAD

**Issue by Yenya** [haproxy/haproxy#2711]
### Detailed Description of the Problem

During the traffic spike, haproxy crashed with all threads having stuck=1 in the crash report (attached below). The traffic spike is probably from some bot, as the client IP addresses belong to the Tencent cloud.

### Expected Behavior

Haproxy should not crash, even on load spikes :-)

### Steps to Reproduce the Behavior

Sorry, I don't know how to reproduce this manually. But the same problem appears in my log files from time to time. It is not overly critical, as crashed haproxy gets restarted by its systemd unit.

### What is your configuration?

```haproxy
global
	maxconn 100000
	# https://www.haproxy.com/documentation/haproxy-configuration-manual/latest/#8
	# log [::1]:514 local1 notice
	log stderr format iso local1 notice
	# chroot /var/empty
	user haproxy
	# group haproxy

	default-path config

	stats timeout 60m
	stats socket /run/haproxy/myservice4.sock mode 660 level admin

	ssl-default-bind-ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256
	ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets
	ssl-server-verify none
	crt-base /somepath

	spread-checks 5

defaults
	timeout connect 60s
	timeout client 7200s
	timeout server 7200s

	mode http

	# log global
	log [::1]:514 local0 info
	option httplog

	default-server ssl check inter 5s downinter 60s
	timeout check 10s

	option forwardfor

frontend myservice4
	bind X.Y.Z.90:80
	bind X.Y.Z.90:443 ssl crt-list myservice.list
	default_backend myservice-servers
	errorfile 503 myservice4-error.http

	http-request redirect code 302 prefix http://my.acme.host if { path -i -m beg /.well-known/acme-challenge/ }

	acl admin_address src ip6prefix::/64 ip4prefix/24
	http-request deny if { path -i -m beg /server/status/haproxy } !admin_address
	stats enable
	stats uri /server/status/haproxy
	stats refresh 10s

	errorfile 403 myservice4-error.http
	http-request deny if { path -i -m beg /server/status/503 }

backend myservice-servers
	balance roundrobin
	stick-table type string len 6 size 16k expire 10m
	stick on cookie(__Host-cookiename)
	option httpchk GET /my/check/url
	http-check expect status 200
	http-check expect string my.service.name

	server backend1 X.Y.Z.1:443 maxconn 120
	server backend2 X.Y.Z.2:443 maxconn 120
	server backend3 X.Y.Z.3:443 maxconn 120
	server backend4 X.Y.Z.4:443 maxconn 120
```

### Output of `haproxy -vv`

```plain
HAProxy version 3.0.4-quictls 2024/09/03 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2029.
Known bugs: http://www.haproxy.org/bugs/bugs-3.0.4.html
Running on: Linux 5.14.0-427.22.1.el9_4.x86_64 #1 SMP PREEMPT_DYNAMIC Sun Jun 23 17:57:52 EDT 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv -DMAX_SESS_STKCTR=12
  OPTIONS = USE_LINUX_TPROXY=1 USE_CRYPT_H=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE2=1
  DEBUG   = 

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 3.1.5+quic 30 Jan 2024
Running on OpenSSL version : OpenSSL 3.1.5+quic 30 Jan 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : no (USE_PCRE2_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.4.1 20231218 (Red Hat 11.4.1-3)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

### Last Outputs and Backtraces

```plain
I have two crash records from two different systems: different FE and BE, different main haproxy process, but both running on the same physical host.

I will attach them as a separate file, as this issue is already over Github limit of 64k characters.
```

### Additional Information

This is haproxy I have compiled myself against the latest QuicTLS release, but I don't use H3/QUIC yet (see https://discourse.haproxy.org/t/http-3-firefox-ns-error-net-interrupt/10176 for details). If needed, I can recompile my haproxy with different options and/or SSL library. The host OS is AlmaLinux 9.

**Comment by Yenya** [comment]
[haproxy-crash.txt](https://github.com/user-attachments/files/16929630/haproxy-crash.txt)

**Comment by einval22** [comment]
Hi Yenya !

Thanks a lot for reporting this and for the traces !

Just to narrow it down, could please recompile with "pure" openssl lib,  which is available in your distribution (1.1.* or 3.* ) in order to confirm, that it does not crash with it, during these spikes.

**Comment by wtarreau** [comment]
I really doubt it would be caused by openssl due to the traces that don't show any sign of it but rather show the queue (pendconn_must_try_again). A few other questions:
- did this start to happen after an update or is this your first attempt at using it ? I'm asking because it died in the queuing system (server maxconn) and we've integrated a fix for a rare bug affecting them between 3.0.3 and 3.0.4.
- since it happens from time to time, and assuming it only started to happen with 3.0.4, would you be interested in reverting commit 68492650d ("BUG/MEDIUM: queue: deal with a rare TOCTOU in assign_server_and_queue()") to try again ? That's the only change to the queue, it *does* contain a looping mechanism (a try again approach for ultra-rare events), but if the condition there would catch by accident another case we'd overlooked that would make it loop forever, that could explain what you're observing.

**Comment by wtarreau** [comment]
In all traces there's a combination of at least one pendconn_add() and several penconn_must_try_again(). The former takes the queue's lock, while the latter takes first the queue lock, and second the proxy's lock. I don't see how that could deadlock, but I'm now wondering if there could be code paths from assign_server_and_queue() where we could land here with the proxy's lock held and block the whole chain.

At first glance since pendconn_add() only locks and immediately unlocks the queue for a tree insertion, it's not the one that blocks the others, thus it's definitely waiting there trying to grab the lock. So it means that it's pendconn_must_try_again() that's not making progress with the queue's lock held. This can happen if the proxy's del_lock is held.

What shocks me is that in the second trace, all threads waiting in pendconn_must_try_again() are waiting at the same line. This means it cannot be the inner lock, so it's necessarily the outer one (q->lock). Since there's a single pendconn_add(), it must be the one holding that lock, but why so long? Or we're facing a rare case of pure serialization maybe? Based on the timers it doesn't look so, since no thread made progress for 2 seconds.

**Comment by wtarreau** [comment]
Note to myself to pursue the analysis tomorrow: nothing proves that these functions do not make any progress. Their callers might actually be looping very fast trying to make progress, and since all of them are waiting on a lock, statistically it's where they're caught. So the theory that maybe we're looping back to the beginning still stands.

**Comment by capflam** [comment]
I don't know if it is relevant by I'm able to reproduce the issue by hacking the code. I do so because I suspect there is an issue with the try-lock on the queue in `process_srv_queue()` since the commit mentioned by Willy. `process_srv_queue()` is called when a stream is closed to unblock queued connections, if any. But if it fails to acquire the queue lock, it leaves. On the other hand, when a connection is queued, if there is no longer stream assigned to the server, it is remove from queue to try again to get a server slot. This test is performed under the queue lock. So we can have this scenario (with a maxconn to 1 to keep things simple) :

1. A first stream is assigned to the server to be processed
2. A new stream is added in the queue because the maxconn is reached on this server
3. In the same time the first stream leaves but it cannot get the queue lock because of 2. So it does nothing
4. Another stream comes and it is added in the queue because there is already a stream waiting in the queue. But because there no stream assigned to the server, it retries. At this stage, we can loop. 

Now I must check with @wtarreau how to workaround this situation and of course if it is a valid scenario.

**Comment by Yenya** [comment]
Hello, all!

I have been replacing ipvs+ldirectord with haproxy in my infrastructure for some time now, and I am still fiddling with the configuration. So unfortunately there are too many variables to put blame on 3.0.4 only.

As for the versions: while I briefly used 3.0.3 on one small system, I recently went from 2.9.7 (also with the latest quictls) directly to 3.0.4 on the rest of my systems. Looking at the logs two months back for occurences of `stuck=1` I have seen this problem only for the last two days, after upgrading to 3.0.4.

Also, I have been experimenting with lowering the "maxconn" parameter to smooth out the load of backend servers and use queuing instead. Before that, there were virtually no queued requests on my haproxy instances.

And one more change: I used stickiness based on a session cookie (see the config file attached to my original report), but it apparently generated huge load spikes and uneven load on the backend servers, so yesterday I commented it out and use pure round-robin scheduling (FWIW, I had to do the same with IPVS many years ago, only based on client IP addresses instead of cookies, the load on workers was too unequal). So I see almost no queuing(*) on my systems now, even with lowered maxconn.

And there were no further haproxy crashes since I disabled stickiness, so I guess the problem might be related to queued requests. I will try to enable stickiness on some of my system to verify that assumption.

(*) in the stats page in the backend section, the third column (Queue/Max) there are zeroes for all individual backend servers, but for some haproxy instances a _non-zero_ number in the black row titled "Backend" at the bottom. Can somebody explain that?

EDIT: I have reenabled stickiness on one haproxy instance, and now the stats page shows non-zero number in the Queue/Max column for some but not all the backend servers, but _zero_ in the last row titled "Backend".

**Comment by wtarreau** [comment]
Many thanks for all these details, they're very useful for us. 2.9 and 3.0 are very close together and I guess that 2.9.7 was not different from 3.0.3 so we could clearly imagine that the loop Christopher seems to have found above since my change is indeed what affects you. We'll discuss it tomorrow. If we can't spot an easy fix, we'll likely revert this fix since the bug it fixes has been there for a while and is much rarer than the one that's the consequence of the fix!

**Comment by capflam** [comment]
@Yenya, could you test the following patch ? It should fix the issue but on our side, it is pretty hard to be sure because we are unable to reproduce the issue.

[0001-BUG-MEDIUM-queue-implement-a-flag-to-check-for-the-d.patch.txt](https://github.com/user-attachments/files/16961324/0001-BUG-MEDIUM-queue-implement-a-flag-to-check-for-the-d.patch.txt)

**Comment by Yenya** [comment]
@capflam: ok, I have compiled 3.0.4 with the above patch and started it on ~15 haproxy instances. So far it works, but it has been only a few minutes. I'll let you know if it crashes.

**Comment by wtarreau** [comment]
@Yenya, any new crash so far ? I'd rather merge the fix quickly in fact ;-)

**Comment by Yenya** [comment]
@wtarreau: no crash so far, but also no crash so far on the unpatched version. I run one patched and one unpatched instance with extremely low "maxconn" on backend servers and stickiness to a session cookie to encourage queueing and per-backend spikes, but none has crashed so far. Also, all other instances (patched and stock 3.0.4, but with standard maxconn and without stickiness, so almost no queueing) are running fine.

So I guess the patch at least does not do any harm. Sorry for a "no news" reply.

**Comment by wtarreau** [comment]
Thank you. You don't have to be sorry for not having problems :-) Since we didn't manage to reproduce it despite trying at different load levels and different maxconn values, it's even possible that it can only trigger at your load with the original maxconn values and at certain moments. I'll merge the patch then. It at least fixes the problem on paper and as you said, it doesn't harm. Many thanks again for your responsiveness in testing!
